tilavarauspalvelu/models/user/actions.py

```diff
@@ -1,11 +1,164 @@
+                billing_first_name=self.user.first_name,
+                billing_last_name=self.user.last_name,
+                billing_email=self.user.email,
+                billing_phone="",
+                billing_address_zip=self._anonymize_string(reservation.billing_address_zip, "99999"),
+                billing_address_city=self._anonymize_string(reservation.billing_address_city),
+                billing_address_street=self._anonymize_string(reservation.billing_address_street),
+                working_memo="",
+                free_of_charge_reason=self._anonymize_string(reservation.free_of_charge_reason, SENSITIVE_RESERVATION),
+                cancel_details=self._anonymize_string(reservation.cancel_details, SENSITIVE_RESERVATION),
+                handling_details=self._anonymize_string(reservation.handling_details, SENSITIVE_RESERVATION),
+            )
+
+        audit_log_ids = LogEntry.objects.get_for_objects(reservations).values_list("id", flat=True)
+        LogEntry.objects.filter(id__in=audit_log_ids).delete()
+
+    def anonymize_user_applications(self) -> None:
+        ApplicationSection.objects.filter(application__user=self.user).update(
+            name=SENSITIVE_APPLICATION,
+        )
+        Person.objects.filter(applications__user=self.user).update(
+            first_name=self.user.first_name,
+            last_name=self.user.last_name,
+            email=self.user.email,
+            phone_number="",
+        )
+        Address.objects.filter(applications__user=self.user).update(
+            post_code="99999",
+            city=ANONYMIZED,
+            city_fi=ANONYMIZED,
+            city_en=ANONYMIZED,
+            city_sv=ANONYMIZED,
+            street_address=ANONYMIZED,
+            street_address_fi=ANONYMIZED,
+            street_address_en=ANONYMIZED,
+            street_address_sv=ANONYMIZED,
+        )
+        Application.objects.filter(user=self.user).update(
+            additional_information=SENSITIVE_APPLICATION,
+            working_memo=SENSITIVE_APPLICATION,
+        )
+
+    def _anonymize_string(self, s: str | None, replacement: str = ANONYMIZED) -> str:
+        return replacement if s and s.strip() else s
+
+    def can_anonymize(self) -> UserAnonymizationInfo:
+        month_ago = timezone.now() - relativedelta(months=1)
+
+        has_open_reservations = (
+            self.user.reservations.filter(end__gte=month_ago)
+            .exclude(
+                state__in=[
+                    ReservationStateChoice.CANCELLED.value,
+                    ReservationStateChoice.DENIED.value,
+                ]
+            )
+            .exists()
+        )
+
+        has_open_applications = (
+            self.user.applications.all()
+            .has_status_in(
+                [
+                    ApplicationStatusChoice.RECEIVED.value,
+                    ApplicationStatusChoice.IN_ALLOCATION.value,
```

Comment:
Come to think of it, `HANDLED` can be added here. Anonymization can well wait until any rounds are in the final state.
(Should the application contain allocations, then the reservation check alone would prevent its anonymization, but if all sections were rejected then it would not -- and it might be confusing handlers to encounter anonymized applications while the application round results have not yet been sent to applicants; even if their sections have been rejected and therefore need not be allocated anything even if the round would be reset in the previous state. It will also normally ensure that a `sent_date` has already been set on anonymized applications.)